aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp

```diff
@@ -0,0 +1,138 @@
+      embedding_rows,
+      embedding_cols +
+          8}; // extra 8 bytes to store FP scale and zero_point per row.
+  size_t output_columns = output_shape[1];
+  constexpr float kEpsilon = 1e-8f;
+
+  // Allocate output packed weights
+  auto output = at::empty(
+      output_shape,
+      weight_contig.options().dtype(at::kByte),
+      weight_contig.suggest_memory_format());
+  auto* output_data = output.data_ptr<uint8_t>();
+
+  for (std::size_t row = 0; row < embedding_rows; ++row) {
+    const float* input_row = weight_data + row * embedding_cols;
+    std::uint8_t* output_row = output_data + row * output_columns;
+    float* output_row_scale_zp =
+        reinterpret_cast<float*>(output_row + embedding_cols);
+
+    float minimum_element =
+        *std::min_element(input_row, input_row + embedding_cols);
+    float maximum_element =
+        *std::max_element(input_row, input_row + embedding_cols);
+    float range = maximum_element - minimum_element;
+
+    output_row_scale_zp[0] = range / 255.0f;
+    output_row_scale_zp[1] = minimum_element;
+    const auto inverse_scale = 255.0f / (range + kEpsilon);
+    for (std::size_t col = 0; col < embedding_cols; ++col) {
+      output_row[col] =
+          lrintf((input_row[col] - minimum_element) * inverse_scale);
+    }
+  }
+  return output;
+}
+
+Tensor qembeddingbag_4bit_prepack(const Tensor& weight) {
+  int64_t embedding_rows = weight.size(0);
+  int64_t embedding_cols = weight.size(1);
+
+  Tensor weight_contig = weight.contiguous(weight.suggest_memory_format());
+
+  const auto weight_data = weight.data_ptr<float>();
+  constexpr int BIT_RATE = 4;
+  constexpr int NUM_ELEM_PER_BYTE = 8 / BIT_RATE;
+  TORCH_CHECK(
+      weight_contig.size(weight.dim() - 1) % NUM_ELEM_PER_BYTE == 0,
+      "FloatToFused4BitRowwiseQuantizedOp only works for the number of "
+      "columns a multiple of 2");
+
+  // The "fused" representation stores the scale and bias with the
+  // row-wise quantized data in one tensor.
+  // Since we represent the scale and bias in 16-bit float, we'll use the
+  // last 4 bytes of each row for scale (2 bytes) and bias (2 bytes).
+  // | ... quantized data ... | scale | bias |
+  // |    number_of_columns   |  2B   |  2B  |
+  std::vector<int64_t> output_shape = {
+      embedding_rows,
+      static_cast<std::int64_t>(
+          (embedding_cols + NUM_ELEM_PER_BYTE - 1) / NUM_ELEM_PER_BYTE +
+          2 * sizeof(at::Half))};
+  auto output = at::empty(
+      output_shape,
+      weight_contig.options().dtype(at::kByte),
+      weight_contig.suggest_memory_format());
+  auto* output_data = output.data_ptr<uint8_t>();
+  const auto output_columns = output.size(output.dim() - 1);
+
+  for (int row = 0; row < embedding_rows; ++row) {
+    const float* input_row = weight_data + row * embedding_cols;
+    std::uint8_t* output_row = output_data + row * output_columns;
+
+    at::Half* output_row_scale_zp = reinterpret_cast<at::Half*>(
```

Comment:
optional readability nit: if this is packed at the end of a row, maybe we can move the code down to be below the weight packing, so the code structure follows the data format?